### PR TITLE
feat: add sha256 checksum extension

### PR DESCRIPTION
**Pull request by kneekey23**
*Description of changes:* Adds sha256 checksum extension on a string for calculations in middleware in the sdk

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: Source/AwsCommonRuntimeKit/Utilities.swift [hunk 1]
```diff
@@ -41,6 +41,21 @@ extension String {
         let byteCursor = aws_byte_cursor_from_buf(output)
         return byteCursor.toData().base64EncodedString()
     }
+
+    public func base64EncodedSha256(allocator: Allocator = defaultAllocator, truncate: Int = 0) -> String? {
```
Comment (graebm) [inline Source/AwsCommonRuntimeKit/Utilities.swift:45]: Could this just return `String` instead of `String?`
I don't think it can ever actually fail. aws_sha256_compute() can only fail if allocation fails (that's fatal now) or the output buffer isn't big enough (you're creating the output buffer, so there's no chance of messing that up)

I understand if you want to stick with `String?` because the other existing `base64EncodedMD5()` function returns `String?`

File: Source/AwsCommonRuntimeKit/Utilities.swift [hunk 2]
```diff
@@ -41,6 +41,19 @@ extension String {
         let byteCursor = aws_byte_cursor_from_buf(output)
         return byteCursor.toData().base64EncodedString()
     }
+
+    public func base64EncodedSha256(allocator: Allocator = defaultAllocator, truncate: Int = 0) -> String {
+        let input: UnsafePointer<aws_byte_cursor> = fromPointer(ptr: self.awsByteCursor)
+        let emptyBuffer: UInt8 = 0
+        let bufferPtr: UnsafeMutablePointer<UInt8> = fromPointer(ptr: emptyBuffer)
+        let buffer = aws_byte_buf(len: 0, buffer: bufferPtr, capacity: 32, allocator: allocator.rawValue)
```
Comment (graebm) [inline Source/AwsCommonRuntimeKit/Utilities.swift:49]: Is it possible to use [AWS_SHA256_LEN](https://github.com/awslabs/aws-c-cal/blob/e364e9a42075447acf0ff9386a83c1ca19955b89/include/aws/cal/hash.h#L12) instead of `32` here? I don't know how accessible C stuff is from Swift

File: Test/AwsCommonRuntimeKitTests/crt/UtilityTests.swift [hunk 3]
```diff
@@ -22,5 +22,19 @@ class UtilityTests: XCTestCase {
         
         XCTAssertEqual(md5, "iB0/3YSo7maijL0IGOgA9g==")
     }
+    
+    func testSha256() throws {
+        let hello = "Hello"
+        let sha256 = hello.base64EncodedSha256()
+        XCTAssertEqual(sha256, "GF+NsyJx/iX1Yab8k4suJkMG7DBO2lGAB9F2SCY4GWk=")
```
Comment (graebm) [inline Test/AwsCommonRuntimeKitTests/crt/UtilityTests.swift:29]: probably worth testing a sha of the empty string "" -> "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855" since it's edge-case-y

**Review by graebm — APPROVED** [review]
fix (or not) and ship

File: Source/AwsCommonRuntimeKit/Utilities.swift (hunk 2 above)
Comment (justinboswell) [reply Source/AwsCommonRuntimeKit/Utilities.swift:49]: It's accessible if it's a constant, not a macro

Comment (kneekey23) [reply Source/AwsCommonRuntimeKit/Utilities.swift:49]: updated :)

File: Test/AwsCommonRuntimeKitTests/crt/UtilityTests.swift (hunk 3 above)
Comment (kneekey23) [reply Test/AwsCommonRuntimeKitTests/crt/UtilityTests.swift:29]: added a test with empty string
